packages/fx-core/src/core/middleware/projectUpgrader.ts

```diff
@@ -0,0 +1,202 @@
+      "local_tenantId",
+      "local_applicationIdUris",
+    ],
+  },
+  {
+    plugin: "fx-resource-bot",
+    secret: "localBotPassword",
+    relatedKeys: ["localBotId", "localObjectId", "local_redirectUri", "bots", "composeExtensions"],
+  },
+];
+
+const solutionContext = {
+  plugin: "solution",
+  relatedKeys: ["localDebugTeamsAppId", "teamsAppTenantId"],
+};
+
+export const ProjectUpgraderMW: Middleware = async (ctx: CoreHookContext, next: NextFunction) => {
+  await upgradeContext(ctx);
+  await next();
+};
+
+export async function upgradeContext(ctx: CoreHookContext): Promise<void> {
+  try {
+    const inputs = ctx.arguments[ctx.arguments.length - 1] as Inputs;
+    if (!inputs.projectPath) {
+      ctx.result = err(NoProjectOpenedError());
+      return;
+    }
+    const projectPathExist = await fs.pathExists(inputs.projectPath);
+    if (!projectPathExist) {
+      ctx.result = err(PathNotExistError(inputs.projectPath));
+      return;
+    }
+    const confFolderPath = path.resolve(inputs.projectPath!, `.${ConfigFolderName}`);
+    const settingsFile = path.resolve(confFolderPath, "settings.json");
+    const projectSettings: ProjectSettings = await fs.readJson(settingsFile);
+    const envName = projectSettings.currentEnv;
+
+    // Read context file.
+    const contextPath = path.resolve(confFolderPath, `env.${envName}.json`);
+    const context = await readContext(contextPath);
+
+    // Update value of specific key in context file to secret pattern.
+    // Return: map of updated values.
+    const updatedKeys = updateContextValue(context);
+    if (!updatedKeys || updatedKeys.size == 0) {
+      // No keys need to be updated, which means the file is up-to-date.
+      // Can quit directly.
+      return;
+    }
+
+    // Some keys updated.
+    // Save the updated context and send log.
+    await saveContext(contextPath, context);
+    const core = ctx.self as FxCore;
+    const logger =
+      core !== undefined && core.tools !== undefined && core.tools.logProvider !== undefined
+        ? core.tools.logProvider
+        : undefined;
+    if (logger) {
+      logger.info(
+        "[core]: context version is too low. Will update context and move some config from env to userdata."
+      );
+    }
+
+    // Read UserData file.
+    const userDataPath = path.resolve(confFolderPath, `${envName}.userdata`);
+    const userData = await readUserData(userDataPath);
+
+    // Merge updatedKeys into UserData.
+    mergeKeysToUserDate(userData, updatedKeys);
+
+    // Save UserData
+    await saveUserData(userDataPath, userData);
+  } catch (error) {
+    ctx.result = err(ContextUpgradeError(error));
+  }
+}
+
+async function readUserData(userDataPath: string): Promise<Record<string, string>> {
+  let dict: Record<string, string>;
+  if (await fs.pathExists(userDataPath)) {
+    const dictContent = await fs.readFile(userDataPath, "UTF-8");
+    dict = deserializeDict(dictContent);
+  } else {
+    dict = {};
```

Comment:
Move the `dict = {}` to declare sentence?
```
let dict: Record<string, string>= {}
```

Reply:
Fix. Thank you for pointing that out. @jayzhang You may update the same code in other middleware.